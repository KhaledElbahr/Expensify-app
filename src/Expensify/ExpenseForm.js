import React, { Component } from 'react'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import { v4 as uuid } from 'uuid'

export default class ExpenseForm extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createAt) : moment(),
      calendarfocused: false,
      title: props.expense ? 'Update' : 'Add',
      error: ''
    }
  }

  getValue = e => e.target.value

  onDescriptionChange = e => {
    const description = this.getValue(e)
    this.setState(() => ({ description }))
  }

  onNoteChange = e => {
    const note = this.getValue(e)
    this.setState(() => ({ note }))
  }

  onAmountChange = e => {
    const amount = this.getValue(e)
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }

  onDateChange = createdAt => this.setState(() => ({ createdAt }))

  onFocusChange = calendarfocused => {
    this.setState(() => ({ calendarfocused }))
  }

  onSubmit = e => {
    e.preventDefault()

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount' }))
    } else {
      this.setState(() => ({ error: '' }))
      this.props.onSubmit({
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount, 10) * 100,
        createAt: this.state.createdAt.valueOf()
      })
    }
  }

  render () {
    return (
      <form className='form' onSubmit={this.onSubmit}>
        {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input
          type='text'
          className='text-input'
          placeholder='Description'
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type='text'
          className='text-input'
          value={this.state.amount}
          onChange={this.onAmountChange}
          placeholder='Amount'
        />
        <div className='input-control'>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarfocused.focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            id={uuid()} // PropTypes.string.isRequired,
            showClearDate={true}
            isOutsideRange={() => false}
          />
        </div>
        <textarea
          value={this.state.note}
          className='textarea'
          onChange={this.onNoteChange}
          placeholder='Add a note for your expense (optional)'
        ></textarea>
        <div className="bototn--group">
          <button className='button'>{this.state.title} Expense</button>
          {
            this.props.expense 
            && 
            <button className="button button--danger" onClick={this.props.onRemove}>Delete Expense</button>
          }
        </div>
      </form>
    )
  }
}
