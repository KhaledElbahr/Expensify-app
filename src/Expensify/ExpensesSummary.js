import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectedExpenses from './../redux/selectors/expenses';
import selectExpensesTotal from './../redux/selectors/expenses-total';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
    const ExpenseWord = expensesCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal).format('$00.00');

    return (
        <div>
            {expensesCount !== 0 ? (
                <>
                <h3>Expenses Summary</h3>
                <p>Viewing {expensesCount} {ExpenseWord} totalling amount {formattedExpensesTotal}</p>
                </>
            ) : ( <p>Add New Expenses</p> )}
        </div>
    )
}

const mapStateToProps = (state) => {
    const expenses =  selectedExpenses(state.expenses, state.filters);
    return {
        expensesCount: expenses.length,
        expensesTotal: selectExpensesTotal(expenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);
