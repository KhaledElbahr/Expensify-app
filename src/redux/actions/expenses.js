import { v4 as uuid } from 'uuid';
import db from './../../firebase/firebase';
import { get, push, ref, update, remove } from "firebase/database";
// actions
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

// Add Expense Data
export const addExpenseData = (expenseData = {}) => {
  return (dispatch, getState) => {
    console.log(getState());
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createAt = 0
    } = expenseData;
    const expense = { description, note, amount, createAt };

    return push(ref(db, `users/${uid}/expenses/`), expense)
    .then(data => {
      dispatch(addExpense({
        id: data.key,
        ...expense       
      }));
    })
    .catch((e) => console.log('This failed', e))
  }
}

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

export const editExpenseData = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return update(ref(db, `users/${uid}/expenses/${id}`), updates)
    .then((snapshot) => {
      dispatch(editExpense(id, updates));
      console.log(`${id} expense is updated successfully`);
    })
    .catch((e) => console.log('This failed', e))
  }
}

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const removeExpenseData = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return remove(ref(db, `users/${uid}/expenses/${id}`))
    .then((snapshot) => {
      dispatch(removeExpense({ id }))
      console.log(`${id} expense is deleted successfully`)
    })
    .catch((e) => console.log('This failed', e))
  }
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

// Set Expense Data
export const setExpensesData = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return get((ref(db, `users/${uid}/expenses/`)))
    .then((snapshot) => {
      if(snapshot.exists()) {
        const expenses = [];
        
        snapshot.forEach(child => {
          expenses.push({
            id: child.key,
            ...child.val()
          })
        })
        dispatch(setExpenses(expenses))
      } else {
        console.log('No data available')
      }
    })
    .catch((e) => console.log('This failed', e))
  }
}
