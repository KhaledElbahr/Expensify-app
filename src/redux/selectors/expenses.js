import moment from 'moment';

// Get Visibile Expenses
export default (expenses, {
    text,
    sortBy,
    startDate,
    endDate
}) => {
    return expenses.filter((expense) => {
        // const endDateMatch = typeof endDate !== 'number' || expense.createAt <= endDate;
        // const startDateMatch = typeof startDate !== 'number' || expense.createAt >= startDate;
        const createdAtMoment = moment(expense.createAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'date') {
            return a.createAt < b.createAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })
}
