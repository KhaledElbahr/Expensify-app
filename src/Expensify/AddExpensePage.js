import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addExpenseData } from './../redux/actions/expenses';
import ExpenseForm from './ExpenseForm';
import { useNavigate } from "react-router-dom";

export class AddExpensePage extends React.Component {
    constructor(props) {
        super(props);
    }

    onSubmit = (expense) => {
        this.props.addExpenseData(expense);
        this.props.navigate('/dashboard');
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="container">
                        <h2 className="page-header__title">Add Expense</h2>
                    </div>
                </div>
                <div className="container">
                    <ExpenseForm onSubmit={this.onSubmit} />
                </div>
            </div>
        )
    }
}

// const AddExpensePage = (props) => {
//     const navigate = useNavigate();

//     return (
//     <div>
//         <h2>Add Expense</h2>
//         <ExpenseForm
//         onSubmit={(expense) => {
//             // props.dispatch(addExpense(expense));
//             props.onSubmit(expense);
//             navigate("/");
//         }}
//         />
//     </div>
// )}

const mapDispatchToProps = (dispatch) => {
    const navigate = useNavigate();
    
    return {
        addExpenseData: (expense) => dispatch(addExpenseData(expense)),
        navigate: (toPath) => navigate(toPath)
    }
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);