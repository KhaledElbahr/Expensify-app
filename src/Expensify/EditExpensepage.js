import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpenseData } from '../redux/actions/expenses';
import { useNavigate } from 'react-router-dom';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);
    }

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.navigate('/dashboard');
    }

    onRemoveExpense = () => {
        this.props.removeExpenseData({ id: this.props.expense.id });
        this.props.navigate('/dashboard');
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="container">
                        <h2 className="page-header__title">Edit Expense</h2>
                    </div>
                </div>
                <div className="container">
                    <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.onRemoveExpense}>Remove Expense</button>
                </div>
            </div> 
        )
    }
}

// const EditExpensePage = (props) => (
//     <div>
//         <ExpenseForm 
//         expense={props.expense}
//         onSubmit={(expense) => {
//             props.dispatch(editExpense(props.expense.id, expense));
//             props.history.push('/');
//         }} 
//         />
//         <button onclick={() => (
//             props.dispatch(removeExpense({ id: props.expense.id }));
//             props.history.push('/');
//         )}>Remove</button>
//     </div>
// );

const mapStateToProps = (state) => {
    return {
        expense: state.expenses.find((expense) => expense.id === window.location.href.split('/')[4])
    }
}

const mapDispatchToProps = (dispatch) => {
    const navigate = useNavigate();
    
    return {
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpenseData: (id) => dispatch(removeExpenseData(id)),
        navigate: (toPath) => navigate(toPath),
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
