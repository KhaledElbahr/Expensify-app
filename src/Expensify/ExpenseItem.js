import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from './../redux/actions/expenses';
import { useNavigate } from 'react-router-dom';

export const ExpenseItem = ({
    id, 
    note, 
    description, 
    amount, 
    createAt,
    dispatch
}) => {
    // navigate = useNavigate();

    return (
        <>
        <li>🗒️ {note}</li>
        <li>🔑 {description}</li>
        <li>📏 {amount}</li>
        <li>⏱️ {createAt}</li>
        <button 
        onClick={(e) => {
            // navigate(`/edit-expense/${id}`);
        }}>Edit</button>
        <button 
        onClick={(e) => {
            dispatch(removeExpense({ id }))
        }}>Remove</button>
    </>
    )
}    

export default connect()(ExpenseItem);