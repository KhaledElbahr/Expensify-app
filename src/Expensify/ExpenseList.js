import React from 'react';
import { connect } from 'react-redux';
import ExpenseItem from './ExpenseItem';
import selectedExpenses from '../redux/selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        <h2>Expense List</h2>
        {props.expenses && props.expenses.length !== 0 ? 
        (<>
            {props.expenses.map((expense) => (
            <ExpenseItem key={expense.id} {...expense} /> 
            ))}
        </>) : (<p>There is no Expenses!!</p>)
        }
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectedExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);