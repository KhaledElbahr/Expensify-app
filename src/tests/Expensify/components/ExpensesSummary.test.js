import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ExpensesSummary } from './../../../Expensify/ExpensesSummary';
import expenses from './../../fixtures/expenes';

test('should render ExpensesSummary with 1 expense correctly!', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={235} />);
    expect(toJson(wrapper)).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses correctly!', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={23} expensesTotal={235587558958} />);
    expect(toJson(wrapper)).toMatchSnapshot();
});