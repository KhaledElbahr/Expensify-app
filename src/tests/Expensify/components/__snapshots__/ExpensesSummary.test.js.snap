// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with 1 expense correctly! 1`] = `
<div>
  <h3>
    Expenses Summary
  </h3>
  <p>
    Viewing 
    1
     
    expense
     totalling amount 
    $235.00
  </p>
</div>
`;

exports[`should render ExpensesSummary with multiple expenses correctly! 1`] = `
<div>
  <h3>
    Expenses Summary
  </h3>
  <p>
    Viewing 
    23
     
    expenses
     totalling amount 
    $235587558958.00
  </p>
</div>
`;
